delta/app/src/main/scala/ch/epfl/bluebrain/nexus/delta/routes/ResourcesPracticeRoutes.scala

```diff
@@ -0,0 +1,162 @@
+import ch.epfl.bluebrain.nexus.delta.sdk.permissions.Permissions.resources.{write => Write}
+import ch.epfl.bluebrain.nexus.delta.sdk.resources.NexusSource.DecodingOption
+import ch.epfl.bluebrain.nexus.delta.sdk.resources.model.ResourceRejection
+import ch.epfl.bluebrain.nexus.delta.sdk.resources.{NexusSource, ResourcesPractice}
+import ch.epfl.bluebrain.nexus.delta.sdk.schemas.Schemas
+import ch.epfl.bluebrain.nexus.delta.sdk.schemas.model.SchemaRejection
+import ch.epfl.bluebrain.nexus.delta.sourcing.model.ProjectRef
+import io.circe.generic.extras.Configuration
+import io.circe.generic.extras.semiauto.deriveConfiguredDecoder
+import io.circe.{Decoder, Json}
+import monix.bio.IO
+import monix.execution.Scheduler
+
+import scala.annotation.nowarn
+
+/**
+  * The resource practice routes allowing to do read-only operations on resource
+  */
+final class ResourcesPracticeRoutes(
+    identities: Identities,
+    aclCheck: AclCheck,
+    generateSchema: GenerateSchema,
+    resourcesPractice: ResourcesPractice,
+    schemeDirectives: DeltaSchemeDirectives
+)(implicit
+    baseUri: BaseUri,
+    s: Scheduler,
+    cr: RemoteContextResolution,
+    ordering: JsonKeyOrdering,
+    decodingOption: DecodingOption
+) extends AuthDirectives(identities, aclCheck)
+    with CirceUnmarshalling
+    with RdfMarshalling {
+
+  import schemeDirectives._
+  def routes: Route =
+    baseUriPrefix(baseUri.prefix) {
+      concat(validateRoute, practiceRoute)
+    }
+
+  private def validateRoute: Route =
+    pathPrefix("resources") {
+      extractCaller { implicit caller =>
+        resolveProjectRef.apply { ref =>
+          authorizeFor(ref, Write).apply {
+            (get & idSegment & idSegmentRef & pathPrefix("validate") & pathEndOrSingleSlash) { (schema, id) =>
+              val schemaOpt = underscoreToOption(schema)
+              emit(
+                resourcesPractice.validate(id, ref, schemaOpt).leftWiden[ResourceRejection]
+              )
+            }
+          }
+        }
+      }
+    }
+
+  private def practiceRoute: Route =
+    (get & pathPrefix("practice") & pathPrefix("resources")) {
+      extractCaller { implicit caller =>
+        (resolveProjectRef & pathEndOrSingleSlash) { ref =>
+          authorizeFor(ref, Write).apply {
+            (entity(as[GenerationInput])) { input =>
+              generate(ref, input)
+            }
+          }
+        }
+      }
+    }
+
+  // Call the generate method matching the schema input
+  private def generate(project: ProjectRef, input: GenerationInput)(implicit caller: Caller) =
+    input.schema match {
+      case ExistingSchema(schemaId) =>
+        emit(resourcesPractice.generate(project, schemaId, input.resource).flatMap(_.asJson))
+      case NewSchema(schemaSource)  =>
+        emit(
+          generateSchema(project, schemaSource, caller).flatMap { schema =>
+            resourcesPractice.generate(project, schema, input.resource).flatMap(_.asJson)
+          }
+        )
+    }
```

Comment:
Those are 10 lines of code here without complexity where:
* Most of it is related to how we handle the input
* ResourcesPractis(c)e is about generating resources, not schemas

Reply:
I think this function is unrelated to parsing or response-writing so it should be in a service rather than routes, but that's just my opinion I guess

Reply:
> I think this function is unrelated to parsing or response-writing so it should be in a service rather than routes, but that's just my opinion I guess

It is related to parsing as `GenerationInput` only lives in the scope of the routes and reflects the payload submitted to the user.

> Any time the user calls an API they are ostensibly performing a single action. The user is saying they want to generate using a generated schema. Perhaps that affects they way you design the ResourcesPractise, but it shouldn't really be a concern for the routes
We are talking here about a pattern matching and a flatMap operation.
I prefer this solution than changing `ResourcePractise` or including yet another layer